### ISSUE TITLE
Async drawable doesn't show up if it's the only thing in the text view

### ISSUE THREAD

**Issue by Adonai** [noties/Markwon#189]
* **Markwon version**: 4.2.1-SNAPSHOT + GlidePlugin

Async drawable doesn't show up if it's the only thing in the text view.

In the source code of the `ReplacementSpan` there's this suspicious javadoc

```java
   /**
     * Returns the width of the span. Extending classes can set the height of the span by updating
     * attributes of {@link android.graphics.Paint.FontMetricsInt}. If the span covers the whole
     * text, and the height is not set,
     * {@link #draw(Canvas, CharSequence, int, int, float, int, int, int, Paint)} will not be
     * called for the span.
     */
     public abstract int getSize(...)
```

And `AsyncDrawable.initWithKnownDimensions(...)` is only called if `AsyncDrawableSpan.draw(...)` of the parent span was called.

**Comment by noties** [comment]
Hello @Adonai !

Can you please try the same markdown input with `4.2.0` version? And different image loader? I think there is a chance that drawable is not displayed because of missing bounds. 

Also, I could not reproduce it. So, let's try to pin point the issue:
* How your image is represented in the markdown - markdown syntaxt or HTML? 
* Does it have size constraints (in case of HTML)? 
* Do you use custom ImageSizeResolver? 
* What type of image it is - SVG, GIF, static?

As for javadoc... well, AsyncDrawableSpan will set height if it has loaded result. If has none, then height won't be set (and thus `draw` method not be called). But this is OK, if you have not loaded a result yet

**Comment by Adonai** [comment]
I'll try to craft minimal reproducible example of this, thanks for quick response!

**Comment by Adonai** [comment]
Here it is

```kotlin
    val renderer = Markwon.builder(ctx)
        .usePlugin(GlideImagesPlugin.create(ctx))
        .build()
```

then 

```kotlin
    val md = renderer.toMarkdown("![7QOIF2X.jpg](https://i.imgur.com/ot9fKhy.png)")
    messageBody.text = md
    AsyncDrawableScheduler.schedule(messageBody)

```

**Note 1:** First time it works ok. The second, when images are cached by Glide, it breaks.

**Note 2:** If paired with any text on the other line, it works again:

```kotlin
    val md = renderer.toMarkdown("No issue!\n\n![7QOIF2X.jpg](https://i.imgur.com/7QOIF2X.jpg)")
    messageBody.text = md
    AsyncDrawableScheduler.schedule(messageBody)
```

**Comment by Adonai** [comment]
I think this may be the culprit: b55b1f0dccd595a553d9c6cc56ef6e25eff5d99f

**Comment by noties** [comment]
Hello @Adonai !

It might be the culprit but did you test with the previous version? 😄 

I did reproduce your report but with a single change it can be fixed:

```java
messageBody.setText(md, TextView.BufferType.SPANNABLE);
```

But I recommend to use the:

```java
markwon.setMarkdown(messageBody, markdown);
```

which takes care of not only that, but also other things (like measuring ordered lists width, etc)

**Comment by Adonai** [comment]
@noties both changes didn't work for me. Still empty message.
Tried on Android Emulator API 27, and real device, API 29.

> It might be the culprit but did you test with the previous version?

No, I haven't... not sure it's possible to select specific version with `-SNAPSHOT`.

**Comment by noties** [comment]
Weird, it works for me, I will take a closer look.

> No, I haven't... not sure it's possible to select specific version with -SNAPSHOT.
I think it is as simple as changing the library version in your dependencies block:
```groovy
implementation 'io.noties.markwon:core:4.2.0' // or any other version, really
```
Checking against latest released version would suffice here, can you try it?

**Comment by Adonai** [comment]
Hi! I'll be out of town for some time, will surely check once I return. Thanks!

**Comment by LaurentTreguier** [comment]
Hello !

I am seeing what looks like the same issue. For me, downgrading to `4.2.0` fixes it (I've used that version since December without experiencing this), and trying `4.2.1` again brings the problem back.
I see it happening in a `RecyclerView` with a `MarkwonAdapter`; using `createTextViewIsRoot()` to get a simple adapter where every entry is only a `TextView`. (If images are put in their own entries, then it matches this issue)

I'd be happy to help if there is anything I can do/test.

**Comment by noties** [comment]
Hello @LaurentTreguier !

On what API level(s) do you see this behavior?

**Comment by LaurentTreguier** [comment]
This is on API 29, on the emulator

**Comment by noties** [comment]
@LaurentTreguier I still cannot reproduce this issue even on API 29 emulator. Can you please share your `Markwon` configuration and a markdown snippet as a reference?

**Comment by LaurentTreguier** [comment]
I'll put this here to make sure it's not lost in the middle of the code:
I got curious and tried a few things; and the problem goes away as soon as I comment out lines adding a placeholder. Removing the custom `PostPlugin` (posted below) didn't change anything. It looks like something to do with replacing placeholders with the actual image.

Here is the function I use to setup The Markwon instance:
```kotlin
fun Fragment.lazyMarkdown() = lazy {
    requireActivity().let {
        Markwon.builder(it)
            .usePlugin(CorePlugin.create())
            .usePlugin(MovementMethodPlugin.create())
            .usePlugin(PostPlugin.create())
            .usePlugin(StrikethroughPlugin.create())
            .usePlugin(ImagesPlugin.create())
            .usePlugin(GlideImagesPlugin.create(PostGlideStore(it)))
            .build()
    }
}
```

Here is the `PostPlugin`:
```kotlin
class PostPlugin private constructor() : AbstractMarkwonPlugin() {
    override fun configureVisitor(builder: MarkwonVisitor.Builder) {
        builder.on(SoftLineBreak::class.java) { visitor, _ -> visitor.forceNewLine() }
        builder.on(Text::class.java) { visitor, text ->
            visitor.builder().append(
                SpannableString(text.literal).apply {
                    LinkifyCompat.addLinks(this, Linkify.WEB_URLS)
                }
            )
            visitor.visitChildren(text)
        }
    }

    override fun configureConfiguration(builder: MarkwonConfiguration.Builder) {
        builder.imageSizeResolver(object : ImageSizeResolver() {
            override fun resolveImageSize(drawable: AsyncDrawable): Rect {
                val canvasWidth = drawable.lastKnownCanvasWidth
                val factor = canvasWidth.toFloat() / drawable.intrinsicWidth
                return Rect(0, 0, canvasWidth, (drawable.intrinsicHeight * factor).toInt())
            }
        })
    }

    companion object {
        fun create() = PostPlugin()
    }
}
```

And here is the `PostGlideStore` implementation:
```kotlin
class PostGlideStore(private val context: Context) : GlideImagesPlugin.GlideStore {
    override fun load(drawable: AsyncDrawable) = AppGlide.with(context)
        .load(drawable.destination)
        .placeholder( // removing this placeholder fixes the issue
            VectorDrawableCompat.create(
                context.resources,
                R.drawable.ic_image,
                context.theme
            )
        )

    override fun cancel(target: Target<*>) = AppGlide.with(context).clear(target)
}
```

Here is a raw markdown string with which that I have seen the issue, using `MarkwonAdapter`:
```markdown
# Gorillaz - El Mañana
[![undefined](https://img.youtube.com/vi/gs1I8_m4AOM/0.jpg)](https://www.youtube.com/watch?v=gs1I8_m4AOM)

[Verse 1]
Summer don't know me no more
He got mad, that's all
Summer don't know me
He'd just let me low in myself
'Cause I do know love
From you then, just dying

[Hook]
I saw that day
Lost my mind
Lord, I'm fine
Maybe in time, you'll want to be mine

[Verse 2]
Don't stop the bud, when it comes
It's the dawn, you'll see
Money won't get there
Ten years passed tonight, you'll flee
If you do that
I'll be sworn
To find you

[Hook]
I saw that day
Lost my mind
Lord, I'm fine
Maybe in time, you'll want to be mine

[Bridge]
I saw that day
Lost my mind
Lord, I'm fine
Maybe in time, you'll want to be mine

[Outro]
(I saw that day)
(Lost my mind)
(Lord, I’m fine)
Maybe in time, you'll want to be mine
Maybe in time, you'll want to be mine
Maybe in time, you'll want to be mine
Maybe in time, you'll want to be mine
```

**Comment by noties** [comment]
Hello @LaurentTreguier !

Thank you for the info, it was most valuable! I did manager to reproduce the issue and the fix is already in the `develop` branch (`4.2.2-SNAPSHOT` must be published soon after github runs the tests).

So, the problem here might indeed be the [commit](https://github.com/noties/Markwon/commit/b55b1f0dccd595a553d9c6cc56ef6e25eff5d99f) that @Adonai mentioned before. In the commit I tried to reduce the number of invalidations that AyncDrawable goes through. For this to properly to function `AsyncDrawable` must be initialized with _no_ bounds. This way, when result is downloaded` `AsyncDrawable` will be invalidated exactly once. But the placeholder did slip my mind. 

So, when you have a `PlaceHolderProvider` and it returns a Drawable without bounds - `AsyncDrawable` was not showing both (due to reduced number of invalidations exactly). This is already fixed in `develop`.

To have a workaround in current `4.2.1` version you can use a placeholder drawable with bounds set. Which makes sense to do anyway, as without bounds placeholder has no real usage (won't be shown).

Currently placeholder bounds has few rules:
* if it has exact bounds - they will be used (no matter what bounds `AsyncDrawable` result might be shown later). This makes sense for regular image loading (in markdown syntax) and in HTML. HTML images can have explicit size (for example `<img width=100%>`), so when image has explicit size this placeholder will keep its bounds (and won't be scaled)
* if placeholder has no explicit bounds, then its _intrinsic_ bounds will be used (in case of HTML this size will be used in scaling)
* if placeholder has no bounds and no intrinsic bounds, then this placeholder won't be shown

Hope this resolves this issue and clears a bit how placeholders are displayed

**Comment by Adonai** [comment]
Hello, I'm back. Will test this in the following days and close the issue. Thanks for your thorough investigation!

**Comment by noties** [comment]
Let me know if issue persists

**Comment by LaurentTreguier** [comment]
I don't know if @Adonai has had time to test the 4.2.2 version, but on my end, it looks like the issue is still here.
I tested again with the example I gave up above, and also with the image part alone (`[![undefined](https://img.youtube.com/vi/gs1I8_m4AOM/0.jpg)](https://www.youtube.com/watch?v=gs1I8_m4AOM)`); and the image still doesn't appear when using a placeholder in the custom `GlideStore`.

**Comment by noties** [comment]
I can reproduce it, the fix is going to be in the next `4.3.0` version (and upcoming snapshot). The problem is that `Glide` is setting placeholder as a regular result. `AsyncDrawable` receives the result, but as there are no canvas dimensions are available - there are no bounds available, thus span returns `0` for the size. Please note that this is the case when there is exactly one image span in the markdown. If there is at least one other character/span, result would be as OK

**Comment by LaurentTreguier** [comment]
Good news, from my testing today, it looks like this issue is fixed for me; thanks @noties !
